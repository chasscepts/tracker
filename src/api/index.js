import axios from 'axios';

const normalizeError = (err) => {
  if (!err) {
    return { message: 'An unknown error encountered. Please try again.' };
  }
  if (err.response) {
    return { message: err.response.data.message || JSON.stringify(err.response.data) };
  }
  if (err.request) {
    return { message: 'Server is not responding.' };
  }
  if (err.message) {
    return { message: err.message };
  }
  if (typeof err === 'string') {
    return { message: err };
  }
  return { message: 'An unknown error encountered. Please try again.' };
};

const instantiate = (token) => {
  const instance = axios.create({
    baseURL: 'http://localhost:3000/',
    responseType: 'json',
  });
  instance.defaults.headers.common.Authorization = `Bearer ${token}`;

  return instance;
};

/**
 *
 * @param {axios} instance An instance of axios to use for this request
 * @param {string} path relative url
 * @returns Promise that resolves to fetched data when request is successful
 * and rejects with error when request fails
 */
const get = (instance, path) => new Promise((resolve, reject) => {
  instance.get(path)
    .then((res) => resolve(res.data))
    .catch((err) => reject(normalizeError(err)));
});

/**
 *
 * @param {axios} instance An instance of axios to use for this request
 * @param {string} path relative url
 * @param {{ string: any }} data sent in body of this post
 * @returns Promise that resolves to fetched data when request is successful
 * and rejects with error when request fails
 */
const post = (instance, path, data) => new Promise((resolve, reject) => {
  instance.post(path, data)
    .then((res) => resolve(res.data))
    .catch((err) => reject(normalizeError(err)));
});

const api = {
  getGroups: (token) => get(instantiate(token), '/groups'),
  getTasks: (token, start, end) => {
    let path = '/tasks';
    if (start) {
      path = `${path}?start=${start}`;
      if (end) {
        path = `${path}&end=${end}`;
      }
    }

    return get(instantiate(token), path);
  },
  updateTask: (token, title) => post(instantiate(token), '/tasks', { title }),
};

export default api;
