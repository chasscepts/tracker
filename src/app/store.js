import { configureStore } from '@reduxjs/toolkit';
import authReducer from '../reducers/authSlice';
import tasksReducer from '../reducers/tasksSlice';
import timerReducer from '../reducers/timerSlice';

export default configureStore({
  reducer: {
    auth: authReducer,
    tasks: tasksReducer,
    timer: timerReducer,
  },
});
