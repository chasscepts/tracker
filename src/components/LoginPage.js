import { useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Redirect, useLocation } from 'react-router';
import { Link } from 'react-router-dom';
import mb from '../assets/css/MobileInput.module.css';
import {
  selectLoginError, login, selectUser, clearLoginError,
} from '../reducers/authSlice';
import { COLORS } from '../utilities';
import LoadingButton from './LoadingButton';

const styles = {
  container: {
    height: '100%',
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'center',
    padding: '30px',
  },
  innerWrap: {
    position: 'relative',
    padding: '15px',
    border: '1px solid #ddd',
    borderRadius: '8px',
  },
  h2: {
    color: COLORS.primary,
  },
  controls: {
    marginTop: '30px',
    display: 'flex',
    justifyContent: 'flex-end',
  },
  link: {
    color: '#9e0606',
    textDecoration: 'underline',
  },
  error: {
    fontSize: '0.8rem',
    fontWeight: 'bold',
    color: 'red',
    textAlign: 'center',
    padding: '15px 0',
  },
};

export default function LoginPage() {
  const [email, setEmail] = useState('');
  const [password, setPassord] = useState('');
  const [localError, setLocalError] = useState(null);
  const [busy, setBusy] = useState(false);
  const error = useSelector(selectLoginError);
  const user = useSelector(selectUser);
  const state = useLocation();
  const dispatch = useDispatch();

  if (user) {
    return <Redirect to={state?.from || '/'} />;
  }
  if (error) {
    if (busy) {
      setBusy(false);
    }
    if (error !== localError) {
      setLocalError(error);
    }
    setTimeout(() => dispatch(clearLoginError()), 0);
  }

  const handleTextChange = (evt) => {
    const { name, value } = evt.target;
    if (name === 'email') {
      setEmail(value);
    } else if (name === 'password') {
      setPassord(value);
    }
  };

  const handleSubmit = () => {
    if (email && password) {
      setBusy(true);
      dispatch(login(email, password));
    }
  };

  return (
    <div style={styles.container}>
      <div style={styles.innerWrap}>
        <h2 style={styles.h2}>Sign In</h2>
        {localError && <div style={styles.error}>{localError}</div>}
        <input className={mb.text} type="text" name="email" value={email} placeholder="Enter Email" onChange={handleTextChange} />
        <input className={mb.text} type="password" name="password" value={password} placeholder="Enter Password" onChange={handleTextChange} />
        <LoadingButton label="Log In" onClick={handleSubmit} loading={busy} />
        <div style={styles.controls}>
          <Link to="/register" style={styles.link}>Register</Link>
        </div>
      </div>
    </div>
  );
}
