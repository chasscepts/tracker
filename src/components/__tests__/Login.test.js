import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import LoginPage from '../LoginPage';
import store from '../../app/store';
import { fetchGroups, fetchTasks, fetchUser } from '../../api';

jest.mock('../../api');

const groups = [{ id: 1, title: 'Group 1' }];

const tasks = [
  {
    id: 1,
    title: 'Task 1',
    group_id: 1,
    user_id: 1,
    entries: [
      {
        id: 1, task_id: 1, entry_date: '2021-09-21', duration: 0,
      },
    ],
  },
];

function Wrapper() {
  return (
    <Provider store={store}>
      <MemoryRouter>
        <LoginPage />
      </MemoryRouter>
    </Provider>
  );
}

describe('Login', () => {
  it('matches snapshot', () => {
    const tree = renderer
      .create(<Wrapper />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('displays the heading', () => {
    render(<Wrapper />);
    expect(screen.getByText('Sign In')).toBeInTheDocument();
  });

  it('displays the register link', async () => {
    render(<Wrapper />);
    expect(screen.getByText('Register')).toBeInTheDocument();
  });

  it('calls api to fetch user', async () => {
    fetchUser.mockResolvedValue({ user: { email: 'test@example.com', token: '111111111' } });
    fetchGroups.mockResolvedValue(groups);
    fetchTasks.mockResolvedValue(tasks);

    render(<Wrapper />);
    userEvent.type(screen.getByPlaceholderText('Enter Email'), 'test.example.com');
    userEvent.type(screen.getByPlaceholderText('Enter Password'), 'password');
    await userEvent.click(screen.getByText('Log In'));
    await Promise.resolve();
    expect(fetchUser).toHaveBeenCalledTimes(1);
  });
});
