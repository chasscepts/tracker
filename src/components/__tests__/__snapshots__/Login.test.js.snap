// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login matches snapshot 1`] = `
<div
  className="container"
  style={
    Object {
      "alignItems": "center",
      "display": "flex",
      "flexDirection": "column",
      "height": "100%",
      "justifyContent": "center",
      "padding": "30px",
    }
  }
>
  <div
    style={
      Object {
        "border": "1px solid #ddd",
        "borderRadius": "8px",
        "maxWidth": "400px",
        "padding": "30px",
        "position": "relative",
        "width": "100%",
      }
    }
  >
    <h2
      style={
        Object {
          "color": "#62b5e5",
        }
      }
    >
      Sign In
    </h2>
    <input
      className="text"
      name="email"
      onChange={[Function]}
      placeholder="Enter Email"
      type="text"
      value=""
    />
    <input
      className="text"
      name="password"
      onChange={[Function]}
      placeholder="Enter Password"
      type="password"
      value=""
    />
    <button
      onClick={[Function]}
      style={
        Object {
          "backgroundColor": "#62b5e5",
          "borderRadius": "4px",
          "borderWidth": 0,
          "boxShadow": "0 3px 9px rgb(0 0 0 / 5%)",
          "color": "#fff",
          "cursor": "pointer",
          "display": "block",
          "fontSize": "1.4rem",
          "fontWeight": "bold,",
          "lineHeight": 1,
          "marginTop": "30px",
          "padding": "15px",
          "textAlign": "center",
          "transition": "all 180ms",
          "width": "100%",
        }
      }
      type="button"
    >
      Log In
    </button>
    <div
      style={
        Object {
          "display": "flex",
          "justifyContent": "flex-end",
          "marginTop": "30px",
        }
      }
    >
      <a
        href="/register"
        onClick={[Function]}
        style={
          Object {
            "color": "#9e0606",
            "textDecoration": "underline",
          }
        }
      >
        Register
      </a>
    </div>
  </div>
</div>
`;
